Enhanced GraphQL queries without requiring database conversion

barnstee opened the pull request:
Also NOT requiring additional GraphQL.entityframework NuGet or additional GraphQL controller or additional tables
Features:
* AddressSpace, Category and Contributor GraphQL queries
* Performance optimized via Entity Framework Core
* Pagination support
* Filtering/Where Expression support
* Several bug fixes in the GraphQL client that I came across while implementing

BoBiene commented:
@barnstee did you make any changes to the configuration of the build action?
currently the build result is not shown in the pr.

barnstee commented:
Works for me:

***@***.***

Erich Barnstedt
Chief Architect Standards & Consortia, Azure IoT
European Microsoft Innovation Center GmbH | Gewürzmühlstrasse 11 | 80539 Munich | Germany | Geschäftsführer/General Managers: Keith Dolliver, Benjamin O. Orndorff   | Amtsgericht Aachen, HRB 12066

From: Bo Biene ***@***.***>
Sent: Freitag, 25. März 2022 10:56
To: OPCFoundation/UA-CloudLibrary ***@***.***>
Cc: Erich Barnstedt ***@***.***>; Mention ***@***.***>
Subject: Re: [OPCFoundation/UA-CloudLibrary] Enhanced GraphQL queries without requiring database conversion (PR #93)


@barnstee<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbarnstee&data=04%7C01%7Cerichb%40microsoft.com%7C8b3989bfc8464fd3c33608da0e45ad0f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637837989681760892%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=t4q6DCdQwVsAZK7NjfD8IDZmhsO%2FbGXaDWJVyl3nXFw%3D&reserved=0> did you make any changes to the configuration of the build action?
currently the build result is not shown in the pr.

-
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOPCFoundation%2FUA-CloudLibrary%2Fpull%2F93%23issuecomment-1078844851&data=04%7C01%7Cerichb%40microsoft.com%7C8b3989bfc8464fd3c33608da0e45ad0f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637837989681760892%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=dC2KR%2FRv96tVOBllzLjoidxsVxcEXOg%2BYElAs29huKE%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAB6HU6W5USWVR6EAYZ6SADDVBWETJANCNFSM5RTUUFPA&data=04%7C01%7Cerichb%40microsoft.com%7C8b3989bfc8464fd3c33608da0e45ad0f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637837989681760892%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=sb3Z8ThmMyt5%2F%2F6bMeP6qu2H%2BIc%2BIUqwrJ%2BaC4YO6wQ%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

barnstee commented:
> @barnstee did you make any changes to the configuration of the build action? currently the build result is not shown in the pr.

![image](https://user-images.githubusercontent.com/8157818/160099409-ffcf3809-a24b-4751-b949-20985a5e267a.png)

BoBiene commented:
@barnstee the docker compose run is gone with this pullrequest.

barnstee commented:
> @barnstee the docker compose run is gone with this pullrequest.

It is now in a separate solution as it was slowing down the compile time significantly.

BoBiene commented:
Schema compare between of graphql schema between this pullrequest and from pullrequest #56
https://www.diffchecker.com/NIHaG3TA

primary difference if found between this and #56

**#56 supports:**
1. strongly typed in the Where-Expression
2. ordering
3. enhanced pagination based on cursors (https://graphql.org/learn/pagination/)
4. datetimes (e.g. publication or modification date)
5. Non-Null of a lot of attributes

BoBiene commented:
> > @barnstee the docker compose run is gone with this pullrequest.
> 
> It is now in a separate solution as it was slowing down the compile time significantly.

You can add multiple configurations in a solution and uncheck projects if you don't need them, there should be no need to put them in a separate solution:
![image](https://user-images.githubusercontent.com/23037659/160557361-c0fc9b01-a125-4f0a-b8f6-8b4bf70d306c.png)

BoBiene commented:
> Schema compare of graphql schema between this pullrequest and #56 https://www.diffchecker.com/NIHaG3TA
> 
> primary difference if found between this and #56
> 
> **#56 supports:**
> 
> 1. strongly typed in the Where-Expression
> 2. ordering
> 3. enhanced pagination based on cursors (https://graphql.org/learn/pagination/)
> 4. datetimes (e.g. publication or modification date)
> 5. Non-Null of a lot of attributes

i think the limitations regarding the pagination is OK.
@barnstee can you implement the other points?

barnstee commented:
> > Schema compare of graphql schema between this pullrequest and #56 https://www.diffchecker.com/NIHaG3TA
> > primary difference if found between this and #56
> > **#56 supports:**
> > 
> > 1. strongly typed in the Where-Expression
> > 2. ordering
> > 3. enhanced pagination based on cursors (https://graphql.org/learn/pagination/)
> > 4. datetimes (e.g. publication or modification date)
> > 5. Non-Null of a lot of attributes
> 
> i think the limitations regarding the pagination is OK. @barnstee can you implement the other points?

Yes, if you help me to figure out how? :-)

barnstee commented:
> > > > Schema compare of graphql schema between this pullrequest and #56 https://www.diffchecker.com/NIHaG3TA
> > > > primary difference if found between this and #56
> > > > **#56 supports:**
> > > > 
> > > > 1. strongly typed in the Where-Expression
> > > > 2. ordering
> > > > 3. enhanced pagination based on cursors (https://graphql.org/learn/pagination/)
> > > > 4. datetimes (e.g. publication or modification date)
> > > > 5. Non-Null of a lot of attributes
> > > 
> > > 
> > > i think the limitations regarding the pagination is OK. @barnstee can you implement the other points?
> > 
> > 
> > Yes, if you help me to figure out how? :-)
> 
> 1. this is probably the hardest part; this is where the Entity-Framework-Mapper comes with its power ;)
> 2. I think here is just an optional OrderBy struct as input-type needed
> 3. keep it with skip & take
> 4. add the datetime scalar type
> 5. add the "non null"
>    ![image](https://user-images.githubusercontent.com/23037659/160640556-06a1ce5a-d3ca-468e-85fe-c8d1a1eb9758.png)

1. I know. Through the Client Library, we have a strongly typed where expression, through.
2. Added.
3. As you have previously mentioned, I also think pagination is OK the way it is.
4. Added.
5. Added.

MarkusHorstmann commented:
I'm still ramping upon the codebase, and have not been in the loop on prior discussions, so take this with a grain of salt.

1. What use case exists for the objectType, dataType queries (in their current form)? I'd have expected at least the node id and namespace uri somewhere, and the ability to filter by nodeset.  Also could simplify by returning all types in the same query with a node type indicating if it's objectType/dataType etc. or is the plan to expose the full information for each type at some point in the future?
2. addressspacetype in the top-level query should be addressSpace, and should expose namespace uri (or a list of URIs, one for each Model in the nodeset). The nodeset_id is also not exposed  (making objectType etc. less useful).
3. nodesetType should also have a namespace Uri and version field.
4. Should pagination use "first" instead of "limit"? a bit more in line with the Relay conventions (https://relay.dev/graphql/connections.htm)
5. Are you familiar with the HotChocolate GraphQL Server library (https://www.nuget.org/packages/HotChocolate)? I've used it in a past project and it seems simpler to use than GraphQL.Net (POCO-base schema/models), yet more flexible and feature rich (pagination, ordering almost "built in"). Has strong adoption and community as well.

barnstee commented:
Thanks Markus for your review, answers below:

> 1. What use case exists for the objectType, dataType queries (in their current form)? I'd have expected at least the node id and namespace uri somewhere, and the ability to filter by nodeset.  Also could simplify by returning all types in the same query with a node type indicating if it's objectType/dataType etc. or is the plan to expose the full information for each type at some point in the future?
[erichb] Yes, we also realized that they are of limited use and the reason we introduced the addressspace query. I will add nodesetid and namespace uri.


> 2. addressspacetype in the top-level query should be addressSpace, and should expose namespace uri (or a list of URIs, one for each Model in the nodeset). The nodeset_id is also not exposed  (making objectType etc. less useful).
[erichb] I will add these two fields and rename the query.

> 3. nodesetType should also have a namespace Uri and version field.
[erichb] Good point, I will add.

> 4. Should pagination use "first" instead of "limit"? a bit more in line with the Relay conventions (https://relay.dev/graphql/connections.htm)
[erichb] Yes, I thought long about how to do this and followed the advice here: https://graphql.org/learn/pagination

> 5. Are you familiar with the HotChocolate GraphQL Server library (https://www.nuget.org/packages/HotChocolate)? I've used it in a past project and it seems simpler to use than GraphQL.Net (POCO-base schema/models), yet more flexible and feature rich (pagination, ordering almost "built in"). Has strong adoption and community as well.
[erichb] I was not aware of this one and simply picked the one from NuGet with the most downloads (which is GraphQL.net).

> 6. Not really specific to this PR/GraphQL: Would it make sense to expose the metadata as a model instead of name value pairs?
[erichb] That is effectively what we're doing with the addressspace query.

MarkusHorstmann commented on UACloudLibraryServer/GraphQL/GraphTypes/AddressSpaceType.cs:
```diff
@@ -36,7 +36,6 @@ public class AddressSpaceType : ObjectGraphType<AddressSpace>
         public AddressSpaceType()
         {
             Field(e => e.Title, type: typeof(NonNullGraphType<StringGraphType>));
```
Should the AddressSpace also have an "identifier" (or id) field?

barnstee replied:
An identifier is only associated with a nodeset file to uniquely identify it. The address space query just wraps it and adds metadata to it.

MarkusHorstmann replied:
In the GraphQL world the AddresSpace is an entity and should have an id, per convention. But let's resolve outside of this PR.

barnstee commented on UACloudLibraryServer/GraphQL/GraphTypes/NodesetType.cs:
```diff
@@ -31,12 +31,15 @@ namespace UACloudLibrary
     {
         public NodesetType()
         {
+            Field(e => e.Identifier, type: typeof(UIntGraphType));
```
we called it "identifier" in the spec and this is why I wanted to be consistent.

MarkusHorstmann replied:
It's a GraphQL standard/convention. But let's not block the PR over this.

MarkusHorstmann commented:
> > 1. What use case exists for the objectType, dataType queries (in their current form)? I'd have expected at least the node id and namespace uri somewhere, and the ability to filter by nodeset.  Also could simplify by returning all types in the same query with a node type indicating if it's objectType/dataType etc. or is the plan to expose the full information for each type at some point in the future?
> >    [erichb] Yes, we also realized that they are of limited use and the reason we introduced the addressspace query. Nodesetid and namespace uri are already part of this query, though.
> > ...

Got it. It might make sense to put these under addressSpace/nodeset? Then they are naturally scoped instead of returning all nodes from all nodesets.

MarkusHorstmann commented on UACloudLibraryServer/GraphQL/UaCloudLibQuery.cs:
```diff
@@ -29,37 +29,79 @@
-                "variabletype",
-                resolve: context => cloudLibRepo.GetVariableTypes()
+            Field<ListGraphType<AddressSpaceType>>(
+                "addressspace",
```
Should this be addressSpace (camel cased)?

barnstee replied:
Do you know what the GraphQL convention is? I thought everything should be lowercase?

jwise-cesmii replied:
I think its a field name, so it should be camelCase:
https://docs.devland.is/technical-overview/api-design-guide/graphql-naming-conventions

MarkusHorstmann replied:
orderBy should also be camel cased, and there are a few typos (oderby, missing "r").

Sorry these are trickling in: I just opened the schema in GraphIQl Explorer, instead of playground, and it makes the schema/query more approachable (to try it just add `app.UseGraphQLGraphiQL();` to startup.cs, and a GraphQL.Server.Ui.GraphiQL nuget ref).

MarkusHorstmann replied:
@barnstee agreed to fix before merge. Resolving.

barnstee commented:
> 

no, nodeset is the actual file while the types are the parsed UA types from the file, hence they should be seperate.

MarkusHorstmann commented:
> > 
> 
> no, nodeset is the actual file while the types are the parsed UA types from the file, hence they should be seperate.

Agreed, it's a logical vs. physical view. Might be worth creating a hierarchy (in GraphQL) using namespace Uri? Probably not for this PR, though.

barnstee commented:
> > > 
> > 
> > 
> > no, nodeset is the actual file while the types are the parsed UA types from the file, hence they should be seperate.
> 
> Agreed, it's a logical vs. physical view. Might be worth creating a hierarchy (in GraphQL) using namespace Uri? Probably not for this PR, though.

We can do that, but I will need your help with that and (I agree) not for this PR. If it is just the camelCase I need to fix, then please approve and I will fix this before my merge tomorrow.
